records/2024-04-11.2FA-magic-link.md

```diff
@@ -0,0 +1,170 @@
+
+In the context of Notify, magic links can be adapted to serve as a simpler second step in the two-factor authentication process. After inputting their username and password, the user checks their email and clicks the magic link to complete the login process. With this approach we can make the login process simpler and more accessible by reducing the number of steps required by the user to log in, thus making the process less obstructive and feel more natural.
+
+
+```mermaid
+
+flowchart LR
+	subgraph Notify Tab
+        id1(Enter username and password)
+        id4(User is logged into Notify)
+    end
+    subgraph Email Inbox Tab
+        id2(Find the magic link)-->id3(Click the link)
+    end
+    id1-->id2
+    id3-->id4
+    id4-->id5((X))
+
+```
+
+## Security Considerations
+While the addition of magic links provides users with a less cumbersome, more accessible and convenient login experience there are a number of security implications to leveraging them.
+
+### Session cookie leaks via Cross-Site-Scripting
+Once a user inputs their username and password, a partially authenticated session cookie is created that could, theoretically, be used by an attacker to gain unauthorized access to Notify. For this to succeed the attacker needs:
+
+1. Access to a user's partially authenticated cookie.
+2. Access to  the user's email inbox to obtain their magic link.
+   * Man-in-the-middle attack to obtain the user's magic link.
+
+
+#### Mitigations
+* CSRF protections buit into Flask, and which are already enabled in our application, configure the `Set-Cookie` header with the `HttpOnly` and `Secure` options, preventing Javascript from reading the contents of the session cookie, and limits the cookie(s) to HTTPS traffic only
+
+* The `Strict-Transport-Security` header is enabled site wide, converting all requests to HTTPS, thus mitigating man-in-the-middle attacks.
+
+* Introduce a process for checking that the `Origin` header matches the target origin when a magic link is followed. As our application is distributed and lives behind a load balancer, making each node aware of it's target origin could prove to be non-trivial task and/or have unintended consequences.
+
+See the [Fallback Mitigation section](#Fallback-Mitigation---An-over-engineered-"last-resort"-solution)
+
+### Email Phishing
+An attacker could tamper with the 2FA email we send to the user and forward them to a phishing site that prompts them to re-input their username and password. If the user is not aware of this type of risk they could unintentionally provide an attacker with their log in credentials.
+
+In this instance, the attacker now has a means to create their own partially authenticated session cookie, and would only need access to the magic link. If they have successfully tampered with the 2FA email, we can assume the attacker already has the means to obtain the magic link, even without direct access to the user's inbox.
+
```

Comment:
After you enter your password and get to a "we've sent you a link to click on" page, we could emphasize that if that link brings them to a page that asks for their password then it's a phishing attempt and they should immediately close the page (and maybe forward the email to us). This might help mitigate the risk.

Reply:
I like this idea, especially if we have the option to make it really stand out. My only worry for our existing users is possibly missing out on the added guidance if we put it on that "we've sent you a link" page. I have personally got into the habit of clicking sign-in and immediately swapping to my inbox to fetch the 2FA code. 
Another option would be to place the guidance directly on the sign-in page for short period of time after release to highlight the changes to the sign in process and a blurb about "staying safe".